src/common/dout.h

```diff
@@ -43,14 +43,48 @@ class DoutPrefixProvider {
   virtual ~DoutPrefixProvider() {}
 };
 
+// helpers
+namespace ceph {
+namespace dout {
+
+template<typename T>
+struct dynamic_marker_t {
+  T value;
+  operator T() const { return value; }
+};
+
+template<typename T>
+dynamic_marker_t<T> need_dynamic(T&& t) {
+  return dynamic_marker_t<T>{ std::forward<T>(t) };
+}
+
+template<typename T>
+struct is_dynamic : public std::false_type {};
+
+template<typename T>
+struct is_dynamic<dynamic_marker_t<T>> : public std::true_type {};
+
+} // dout
+} // ceph
+
 // generic macros
 #define dout_prefix *_dout
 
 #define dout_impl(cct, sub, v)						\
   do {									\
-  if (cct->_conf->subsys.should_gather(sub, v)) {			\
+  const bool should_gather = [&](const auto cctX) {			\
+    if constexpr (ceph::dout::is_dynamic<decltype(sub)>::value) {	\
+      return cctX->_conf->subsys.should_gather(sub, v);			\
+    } else {								\
+      return cctX->_conf->subsys.template should_gather<sub>(v);	\
+    }									\
+  }(cct);								\
+									\
+  if (unlikely(should_gather)) {					\
+    [&]() __attribute__((noinline,cold)) {				\
```

Comment:
This can be controversial as we're using compiler-specific extensions here. However, the befit seems to outweigh the costs. Please take a look on [a comparison](https://gist.github.com/rzarzynski/06b985d6a8340ce50f6f246e5b3a7175) of `KernelDevice::read()` translation before and after the change.

TODO: comment the intention behind the lambda and `noinline,cold`.

TODO: introduce a macro-based abstraction?